src/lib/repositories/impl_oauth2/oauth2_repository_impl.py

```diff
@@ -0,0 +1,257 @@
+                    refresh_token.token,
+                    self.env_config.secret_key,
+                    algorithms="HS256",
+                )
+                new_access_token = self._refresh_token(
+                    decoded_refresh_token, refresh_token.id
+                )
+                return new_access_token
+
+            except jwt.ExpiredSignatureError:
+                self._delete_refresh_token(refresh_token.token)
+                return None
+
+        except jwt.InvalidSignatureError:
+            return None
+        except jwt.DecodeError:
+            return None
+
+    def _refresh_token(self, decoded_refresh_token, refresh_token_id):
+        client = self._get_client_by_id(decoded_refresh_token.get("client_id"))
+
+        if decoded_refresh_token.get("user_name"):
+
+            new_access_token = create_token(
+                {
+                    "user_name": decoded_refresh_token.get("user_name"),
+                    "name": decoded_refresh_token.get("name"),
+                    "last_name": decoded_refresh_token.get("last_name"),
+                    "role": decoded_refresh_token.get("role"),
+                },
+                client.access_token_expiration_time,
+                self.env_config.secret_key,
+                algorithm="HS256",
+            )
+            self._add_access_token(refresh_token_id, new_access_token)
+            return {"new_access_token": new_access_token}
+
+        new_access_token = create_token(
+            {"scope": decoded_refresh_token.get("scope")},
+            client.access_token_expiration_time,
+            self.env_config.secret_key,
+            algorithm="HS256",
+        )
+        self._add_access_token(refresh_token_id, new_access_token)
+        return {"new_access_token": new_access_token}
+
+    def _client_credential_login(self, client):
+        client_scope = self._get_scope_by_app_client_id(client.id)
+
+        refresh_token = create_token(
+            {"scope": client_scope.scope, "client_id": client.id},
+            client.refresh_token_expiration_time,
+            self.env_config.secret_key,
+            algorithm="HS256",
+        )
+        self._add_refresh_token(refresh_token)
+        refresh_token_with_id = self._get_refresh_token_by_token(refresh_token)
+
+        access_token = create_token(
+            {"scope": client_scope.scope},
+            client.access_token_expiration_time,
+            self.env_config.secret_key,
+            algorithm="HS256",
+        )
+        self._add_access_token(refresh_token_with_id.id, access_token)
+
+        return {"access_token": access_token, "refresh_token": refresh_token}
+
+    def _client_user_credential_login(self, client, user_name, password):
+        user = self._get_user_by_user_name(user_name)
+
+        if user:
+            if bcrypt.checkpw(password.encode("utf-8"), user.password.encode("utf-8")):
+                refresh_token = create_token(
```

Comment:
Creating the token seems very repetitive in the code

Reply:
No. just 1 access token and 1 refresh token for client login and 1 access token and 1 refresh token for user login, at the end just 4 token will be created.